app/src/main/java/de/westnordost/streetcomplete/quests/amenity_indoor/AddIsAmenityIndoor.kt

```diff
@@ -0,0 +1,114 @@
+package de.westnordost.streetcomplete.quests.amenity_indoor
+
+import de.westnordost.osmfeatures.FeatureDictionary
+import de.westnordost.streetcomplete.R
+import de.westnordost.streetcomplete.data.elementfilter.toElementFilterExpression
+import de.westnordost.streetcomplete.data.osm.geometry.ElementGeometry
+import de.westnordost.streetcomplete.data.osm.geometry.ElementPolygonsGeometry
+import de.westnordost.streetcomplete.data.osm.mapdata.Element
+import de.westnordost.streetcomplete.data.osm.mapdata.MapDataWithGeometry
+import de.westnordost.streetcomplete.data.osm.osmquests.OsmElementQuestType
+import de.westnordost.streetcomplete.data.user.achievements.EditTypeAchievement.*
+import de.westnordost.streetcomplete.osm.Tags
+import de.westnordost.streetcomplete.quests.YesNoQuestForm
+import de.westnordost.streetcomplete.util.ktx.toYesNo
+import de.westnordost.streetcomplete.util.math.LatLonRaster
+import de.westnordost.streetcomplete.util.math.contains
+import de.westnordost.streetcomplete.util.math.isCompletelyInside
+import de.westnordost.streetcomplete.util.math.isInMultipolygon
+import java.util.concurrent.FutureTask
+
+class AddIsAmenityIndoor(private val featureDictionaryFuture: FutureTask<FeatureDictionary>) :
+    OsmElementQuestType<Boolean> {
+
+    private val nodesFilter by lazy {
+        """
+        nodes with
+         (
+         emergency ~ defibrillator|fire_extinguisher
+         or amenity ~ atm|telephone|parcel_locker|luggage_locker|locker|clock|post_box|public_bookcase|give_box|ticket_validator|vending_machine
+         )
+         and access !~ private|no
+         and !indoor and !location and !level and !level:ref
+    """.toElementFilterExpression()
+    }
+
+    /* We only want survey nodes within building outlines. */
+    private val BuildingFilter by lazy {
+        """
+    ways, relations with building
+    """.toElementFilterExpression()
+    }
+
+    override val changesetComment = "Determine whether amenities are inside buildings"
+    override val wikiLink = "Key:indoor"
+
+    //TODO: Generisches Item
+    override val icon = R.drawable.ic_quest_defibrillator
+    override val achievements = listOf(CITIZEN)
+
+    override fun getTitle(tags: Map<String, String>) = R.string.quest_is_amenity_inside_title
+
+    override fun getApplicableElements(mapData: MapDataWithGeometry): Iterable<Element> {
+        val bbox = mapData.boundingBox ?: return listOf()
+        val nodes = mapData.nodes.filter { isApplicableTo(it) }
+        val buildings = mapData.filter { BuildingFilter.matches(it) }.toMutableList()
+
+        val buildingGeometriesById = buildings.associate {
+            it.id to mapData.getGeometry(it.type, it.id) as? ElementPolygonsGeometry
+        }
+
+        val nodesPositions = LatLonRaster(bbox, 0.0005)
+        for (node in nodes) {
+            nodesPositions.insert(node.position)
+        }
+
+        buildings.removeAll { building ->
```

Comment:
A further performance improvement would be possible here.

Currently for every node, you check if it is contained within any building that is near one of these nodes.

You could also reduce the `nodes` beforehand that are not near any building. (add all nodes near buildings to a set and then only iterate through these instead of all `nodes`)

However, probably not worth it because I guess there are not that many "small amenity nodes" looking at the filter you wrote, so not required for this PR to be merged. Just wanted to note this.